src/dashboards/components/DashboardContainer.tsx

```diff
@@ -26,12 +26,61 @@ const {Active} = AutoRefreshStatus
 type ReduxProps = ConnectedProps<typeof connector>
 type Props = ReduxProps
 
+let timer
 const DashboardContainer: FC<Props> = ({autoRefresh, dashboard}) => {
   const dispatch = useDispatch()
 
+  const registerListeners = useCallback(() => {
+    if (timer) {
+      registerStopListeners()
+    }
+    timer = setTimeout(() => {
+      dispatch(resetDashboardAutoRefresh(dashboard))
+      registerStopListeners()
+    }, autoRefresh.inactivityTimeout)
+
+    window.addEventListener('load', registerListeners)
+    document.addEventListener('mousemove', registerListeners)
+    document.addEventListener('keypress', registerListeners)
+  }, [dashboard, autoRefresh.inactivityTimeout])
+
+  const registerStopListeners = useCallback(() => {
+    // Stop all existing timers and deregister everythang
+    if (timer) {
+      clearTimeout(timer)
+      timer = null
+    }
+
+    window.removeEventListener('load', registerListeners)
+    document.removeEventListener('mousemove', registerListeners)
+    document.removeEventListener('keypress', registerListeners)
+  }, [registerListeners])
+
+  useEffect(() => {
+    if (
+      autoRefresh?.status &&
+      autoRefresh.status === AutoRefreshStatus.Active &&
+      autoRefresh.inactivityTimeout > 0
+    ) {
+      registerListeners()
+    } else {
+      registerStopListeners()
+    }
+    return () => {
+      registerStopListeners()
+    }
+  }, [autoRefresh?.status, autoRefresh.inactivityTimeout])
+
+  const stopFunc = useCallback(() => {
+    if (autoRefresh?.duration?.upper <= new Date().toISOString()) {
+      GlobalAutoRefresher.stopPolling()
+      dispatch(resetDashboardAutoRefresh(dashboard))
+    }
+  }, [dashboard, dispatch, autoRefresh?.duration?.upper])
+
   useEffect(() => {
     if (autoRefresh.status === Active) {
-      GlobalAutoRefresher.poll(autoRefresh.interval)
+      GlobalAutoRefresher.poll(autoRefresh, stopFunc)
```

Comment:
i was on a big rant about how you should be able to inspect state in redux about this refresher, then to my dismay saw that i wrote it, so this is fine